language/vm/src/file_format_common.rs

```diff
@@ -31,6 +31,64 @@ impl BinaryConstants {
     pub const TABLE_HEADER_SIZE: u8 = size_of::<u32>() as u8 * 2 + 1;
 }
 
+pub const TABLE_COUNT_MAX: u64 = 127;
+
+pub const SIGNATURE_COUNT_MAX: u64 = 65535;
+pub const SIGNATURE_INDEX_MAX: u64 = SIGNATURE_COUNT_MAX - 1;
+
+pub const TABLE_OFFSET_MAX: u64 = 0xffff_ffff;
+pub const TABLE_SIZE_MAX: u64 = 0xffff_ffff;
+
+pub const ADDRESS_COUNT_MAX: u64 = 65535;
+pub const ADDRESS_INDEX_MAX: u64 = ADDRESS_COUNT_MAX - 1;
+
+pub const IDENTIFIER_COUNT_MAX: u64 = 65536;
+pub const IDENTIFIER_INDEX_MAX: u64 = IDENTIFIER_COUNT_MAX - 1;
+
+pub const MODULE_HANDLE_COUNT_MAX: u64 = 65535;
+pub const MODULE_HANDLE_INDEX_MAX: u64 = MODULE_HANDLE_COUNT_MAX - 1;
+
+pub const STRUCT_HANDLE_COUNT_MAX: u64 = 65535;
+pub const STRUCT_HANDLE_INDEX_MAX: u64 = STRUCT_HANDLE_COUNT_MAX - 1;
+
+pub const STRUCT_DEF_COUNT_MAX: u64 = 65535;
+pub const STRUCT_DEF_INDEX_MAX: u64 = STRUCT_DEF_COUNT_MAX - 1;
+
+pub const FUNCTION_HANDLE_COUNT_MAX: u64 = 65535;
+pub const FUNCTION_HANDLE_INDEX_MAX: u64 = FUNCTION_HANDLE_COUNT_MAX - 1;
+
+pub const FUNCTION_INST_COUNT_MAX: u64 = 65535;
+pub const FUNCTION_INST_INDEX_MAX: u64 = FUNCTION_INST_COUNT_MAX - 1;
+
+pub const IDENTIFIER_SIZE_MAX: u64 = 65535;
+
+pub const CONSTANT_SIZE_MAX: u64 = 65535;
+
+pub const SIGNATURE_SIZE_MAX: u64 = 127;
+
+pub const TYPE_PARAMETER_COUNT_MAX: u64 = 127;
+pub const TYPE_PARAMETER_INDEX_MAX: u64 = TYPE_PARAMETER_COUNT_MAX - 1;
+
+pub const FIELD_COUNT_MAX: u64 = 127;
+pub const FIELD_OFFSET_MAX: u64 = FIELD_COUNT_MAX - 1;
+
+pub const FIELD_HANDLE_COUNT_MAX: u64 = 65535;
+pub const FIELD_HANDLE_INDEX_MAX: u64 = FIELD_HANDLE_COUNT_MAX - 1;
+
+pub const FIELD_INST_COUNT_MAX: u64 = 65535;
+pub const FIELD_INST_INDEX_MAX: u64 = FIELD_INST_COUNT_MAX - 1;
+
+pub const ACQUIRES_COUNT_MAX: u64 = 127;
+
+pub const BYTECODE_COUNT_MAX: u64 = 65535;
+pub const BYTECODE_INDEX_MAX: u64 = BYTECODE_COUNT_MAX - 1;
+
+pub const STRUCT_DEF_INST_COUNT_MAX: u64 = 65535;
+pub const STRUCT_DEF_INST_INDEX_MAX: u64 = STRUCT_DEF_INST_COUNT_MAX - 1;
+
+pub const CONSTANT_COUNT_MAX: u64 = 65535;
+pub const CONSTANT_INDEX_MAX: u64 = CONSTANT_COUNT_MAX - 1;
+
```

Comment:
Note: some of the numbers here are probably not correct. I will take another look at this. Still, please let me know if you spot something really wrong.

Also keep in mind that I haven't figured out how we should deal with the maximum number of elements in each individual table. (For example, `MODULE_HANDLE_COUNT_MAX` is not directly referenced in ser/de.) Tables are packed in a blob and we may just want to set a limit on the size of the blob as a whole.